File: server/mfup/storage.py
"""MFUP/2 SQLite storage layer.

Each session lives in:
    <base_dir>/.incoming.<session_id>/state.sqlite
    <base_dir>/.incoming.<session_id>/payload/...
"""

from __future__ import annotations

import os
import sqlite3
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

from .protocol import NodeKind, NodeStatus, SessionState, ROOT_NODE_ID


# ---------------------------------------------------------------------------
# Schema
# ---------------------------------------------------------------------------

_SCHEMA = """\
CREATE TABLE IF NOT EXISTS sessions (
    session_id   TEXT PRIMARY KEY,
    resume_token TEXT NOT NULL,
    epoch        INTEGER NOT NULL DEFAULT 1,
    state        TEXT NOT NULL DEFAULT 'active',
    target_dir   TEXT NOT NULL DEFAULT '.',
    expires_at   TEXT NOT NULL,
    created_at   TEXT NOT NULL,
    updated_at   TEXT NOT NULL
);

CREATE TABLE IF NOT EXISTS nodes (
    node_id    INTEGER PRIMARY KEY,
    parent_id  INTEGER NOT NULL,
    kind       TEXT NOT NULL,
    name       TEXT NOT NULL,
    size       INTEGER,
    mtime_ms   INTEGER,
    status     TEXT NOT NULL DEFAULT 'open'
);

CREATE TABLE IF NOT EXISTS files (
    node_id          INTEGER PRIMARY KEY REFERENCES nodes(node_id),
    accepted_offset  INTEGER NOT NULL DEFAULT 0,
    final_size       INTEGER,
    checksum_state   BLOB,
    local_tmp_path   TEXT
);

CREATE TABLE IF NOT EXISTS pruned (
    node_id  INTEGER PRIMARY KEY
);

CREATE TABLE IF NOT EXISTS rejected (
    node_id  INTEGER PRIMARY KEY,
    code     TEXT NOT NULL,
    reason   TEXT NOT NULL
);

CREATE TABLE IF NOT EXISTS root_summary (
    id               INTEGER PRIMARY KEY CHECK (id = 1),
    scan_done_units  INTEGER NOT NULL DEFAULT 0,
    scan_est_units   INTEGER NOT NULL DEFAULT 0,
    body_done_bytes  INTEGER NOT NULL DEFAULT 0,
    body_est_bytes   INTEGER NOT NULL DEFAULT 0,
    sealed           INTEGER NOT NULL DEFAULT 0
);

INSERT OR IGNORE INTO root_summary (id) VALUES (1);
"""


class SessionDB:
    """Thin wrapper around a per-session SQLite database."""

    def __init__(self, db_path: Path) -> None:
        self.path = db_path
        self._conn = sqlite3.connect(str(db_path), timeout=5.0)
        self._conn.execute("PRAGMA journal_mode=WAL")
        self._conn.execute("PRAGMA busy_timeout=3000")
        self._conn.execute("PRAGMA synchronous=NORMAL")
        self._conn.executescript(_SCHEMA)
        self._conn.commit()

    @property
    def conn(self) -> sqlite3.Connection:
        return self._conn

    def close(self) -> None:
        self._conn.close()

    # -- session ------------------------------------------------------------

    def init_session(
        self,
        session_id: str,
        resume_token: str,
        expires_at: str,
        target_dir: str = ".",
    ) -> None:
        now = datetime.now(timezone.utc).isoformat()
        self._conn.execute(
            "INSERT OR REPLACE INTO sessions VALUES (?,?,1,?,?,?,?,?)",
            (session_id, resume_token, SessionState.ACTIVE.value, target_dir, expires_at, now, now),
        )
        self._conn.commit()

    def get_target_dir(self) -> str:
        cur = self._conn.execute("SELECT target_dir FROM sessions LIMIT 1")
        row = cur.fetchone()
        return row[0] if row else "."

    def get_session(self) -> Optional[sqlite3.Row]:
        self._conn.row_factory = sqlite3.Row
        cur = self._conn.execute("SELECT * FROM sessions LIMIT 1")
        return cur.fetchone()

    def set_state(self, state: SessionState) -> None:
        now = datetime.now(timezone.utc).isoformat()
        self._conn.execute(
            "UPDATE sessions SET state=?, updated_at=?", (state.value, now)
        )
        self._conn.commit()

    def increment_epoch(self) -> int:
        self._conn.execute("UPDATE sessions SET epoch = epoch + 1, updated_at = ?",
                           (datetime.now(timezone.utc).isoformat(),))
        self._conn.commit()
        cur = self._conn.execute("SELECT epoch FROM sessions LIMIT 1")
        return cur.fetchone()[0]

    def get_epoch(self) -> int:
        cur = self._conn.execute("SELECT epoch FROM sessions LIMIT 1")
        row = cur.fetchone()
        return row[0] if row else 0

    def get_state(self) -> SessionState:
        cur = self._conn.execute("SELECT state FROM sessions LIMIT 1")
        row = cur.fetchone()
        return SessionState(row[0]) if row else SessionState.ACTIVE

    def get_expires_at(self) -> str:
        cur = self._conn.execute("SELECT expires_at FROM sessions LIMIT 1")
        return cur.fetchone()[0]

    def set_expires_at(self, expires_at: str) -> None:
        self._conn.execute(
            "UPDATE sessions SET expires_at=?, updated_at=?",
            (expires_at, datetime.now(timezone.utc).isoformat()),
        )
        self._conn.commit()

    # -- nodes --------------------------------------------------------------

    def upsert_node(
        self,
        node_id: int,
        parent_id: int,
        kind: NodeKind,
        name: str,
        size: Optional[int] = None,
        mtime_ms: Optional[int] = None,
    ) -> None:
        kind_str = "dir" if kind == NodeKind.DIR else "file"
        self._conn.execute(
            "INSERT OR REPLACE INTO nodes VALUES (?,?,?,?,?,?,?)",
            (node_id, parent_id, kind_str, name, size, mtime_ms, NodeStatus.OPEN.value),
        )
        if kind == NodeKind.FILE:
            self._conn.execute(
                "INSERT OR IGNORE INTO files (node_id, accepted_offset) VALUES (?, 0)",
                (node_id,),
            )
        self._conn.commit()

    def get_node(self, node_id: int) -> Optional[sqlite3.Row]:
        self._conn.row_factory = sqlite3.Row
        cur = self._conn.execute("SELECT * FROM nodes WHERE node_id=?", (node_id,))
        return cur.fetchone()

    def set_node_status(self, node_id: int, status: NodeStatus) -> None:
        self._conn.execute(
            "UPDATE nodes SET status=? WHERE node_id=?", (status.value, node_id)
        )
        self._conn.commit()

    # -- files --------------------------------------------------------------

    def get_file(self, node_id: int) -> Optional[sqlite3.Row]:
        self._conn.row_factory = sqlite3.Row
        cur = self._conn.execute("SELECT * FROM files WHERE node_id=?", (node_id,))
        return cur.fetchone()

    def set_accepted_offset(self, node_id: int, offset: int) -> None:
        self._conn.execute(
            "UPDATE files SET accepted_offset=? WHERE node_id=?", (offset, node_id)
        )
        self._conn.commit()

    def set_file_final(self, node_id: int, final_size: int, local_path: str) -> None:
        self._conn.execute(
            "UPDATE files SET final_size=?, local_tmp_path=? WHERE node_id=?",
            (final_size, local_path, node_id),
        )
        self._conn.commit()

    def get_open_files(self) -> list[sqlite3.Row]:
        self._conn.row_factory = sqlite3.Row
        cur = self._conn.execute(
            "SELECT f.*, n.status FROM files f JOIN nodes n ON f.node_id = n.node_id "
            "WHERE n.status IN ('open')"
        )
        return cur.fetchall()

    def get_all_files(self) -> list[sqlite3.Row]:
        self._conn.row_factory = sqlite3.Row
        cur = self._conn.execute(
            "SELECT f.node_id, f.accepted_offset, n.status "
            "FROM files f JOIN nodes n ON f.node_id = n.node_id"
        )
        return cur.fetchall()

    # -- prune / reject -----------------------------------------------------

    def add_pruned(self, node_id: int) -> None:
        self._conn.execute("INSERT OR IGNORE INTO pruned VALUES (?)", (node_id,))
        self._conn.commit()

    def is_pruned(self, node_id: int) -> bool:
        cur = self._conn.execute("SELECT 1 FROM pruned WHERE node_id=?", (node_id,))
        return cur.fetchone() is not None

    def get_pruned_nodes(self) -> list[int]:
        cur = self._conn.execute("SELECT node_id FROM pruned")
        return [r[0] for r in cur.fetchall()]

    def add_rejected(self, node_id: int, code: str, reason: str) -> None:
        self._conn.execute(
            "INSERT OR IGNORE INTO rejected VALUES (?,?,?)", (node_id, code, reason)
        )
        self.set_node_status(node_id, NodeStatus.REJECTED)
        self._conn.commit()

    def is_rejected(self, node_id: int) -> bool:
        cur = self._conn.execute("SELECT 1 FROM rejected WHERE node_id=?", (node_id,))
        return cur.fetchone() is not None

    def get_rejected_files(self) -> list[int]:
        cur = self._conn.execute("SELECT node_id FROM rejected")
        return [r[0] for r in cur.fetchall()]

    # -- root summary -------------------------------------------------------

    def update_root_summary(
        self,
        scan_done: int,
        scan_est: int,
        body_done: int,
        body_est: int,
        sealed: bool,
    ) -> None:
        self._conn.execute(
            "UPDATE root_summary SET scan_done_units=?, scan_est_units=?, "
            "body_done_bytes=?, body_est_bytes=?, sealed=? WHERE id=1",
            (scan_done, scan_est, body_done, body_est, int(sealed)),
        )
        self._conn.commit()

    def get_root_summary(self) -> dict:
        self._conn.row_factory = sqlite3.Row
        cur = self._conn.execute("SELECT * FROM root_summary WHERE id=1")
        row = cur.fetchone()
        if not row:
            return {
                "scan_done_units": 0,
                "scan_est_units": 0,
                "body_done_bytes": 0,
                "body_est_bytes": 0,
                "sealed": False,
            }
        return {
            "scan_done_units": row["scan_done_units"],
            "scan_est_units": row["scan_est_units"],
            "body_done_bytes": row["body_done_bytes"],
            "body_est_bytes": row["body_est_bytes"],
            "sealed": bool(row["sealed"]),
        }

    # -- aggregate stats for COMMIT_OK --------------------------------------

    def count_committed_files(self) -> tuple[int, int]:
        """Return (file_count, total_bytes) for all non-rejected files."""
        cur = self._conn.execute(
            "SELECT COUNT(*), COALESCE(SUM(accepted_offset), 0) "
            "FROM files f JOIN nodes n ON f.node_id = n.node_id "
            "WHERE n.status NOT IN ('rejected', 'pruned')"
        )
        row = cur.fetchone()
        return (row[0], row[1])

    def get_incomplete_files(self) -> list[dict]:
        """Return files where accepted_offset != final_size (incomplete transfers)."""
        self._conn.row_factory = sqlite3.Row
        cur = self._conn.execute(
            "SELECT f.node_id, f.accepted_offset, f.final_size "
            "FROM files f JOIN nodes n ON f.node_id = n.node_id "
            "WHERE n.status NOT IN ('rejected', 'pruned') "
            "  AND (f.final_size IS NULL OR f.accepted_offset != f.final_size)"
        )
        return [dict(row) for row in cur.fetchall()]


# ---------------------------------------------------------------------------
# Staging directory helpers
# ---------------------------------------------------------------------------

DEFAULT_STAGING_PREFIX = ".incoming"


def staging_dir(base_dir: Path, session_id: str, prefix: str = DEFAULT_STAGING_PREFIX) -> Path:
    return base_dir / f"{prefix}.{session_id}"


def ensure_staging(base_dir: Path, session_id: str, prefix: str = DEFAULT_STAGING_PREFIX) -> Path:
    sd = staging_dir(base_dir, session_id, prefix)
    sd.mkdir(parents=True, exist_ok=True)
    (sd / "payload").mkdir(exist_ok=True)
    return sd


def open_session_db(base_dir: Path, session_id: str, prefix: str = DEFAULT_STAGING_PREFIX) -> SessionDB:
    sd = ensure_staging(base_dir, session_id, prefix)
    return SessionDB(sd / "state.sqlite")


def validate_node_name(name: str) -> None:
    """Reject names that could escape the payload directory."""
    if not name:
        raise ValueError("empty node name")
    if name in (".", ".."):
        raise ValueError(f"illegal node name: {name!r}")
    if "/" in name or "\\" in name or "\x00" in name:
        raise ValueError(f"illegal characters in node name: {name!r}")


def resolve_payload_path(base_dir: Path, session_id: str, db: SessionDB, node_id: int, prefix: str = DEFAULT_STAGING_PREFIX) -> Path:
    """Build the filesystem path for a node inside the payload directory.

    Walks parent_id chain in the DB to reconstruct the relative path.
    Validates each name component to prevent path traversal.
    """
    parts: list[str] = []
    cur_id = node_id
    while cur_id != ROOT_NODE_ID:
        node = db.get_node(cur_id)
        if node is None:
            break
        validate_node_name(node["name"])
        parts.append(node["name"])
        cur_id = node["parent_id"]
    parts.reverse()
    sd = staging_dir(base_dir, session_id, prefix)
    result = sd / "payload" / Path(*parts) if parts else sd / "payload"
    # Defense in depth: resolved path must stay within payload dir
    payload_root = sd / "payload"
    if not result.resolve().is_relative_to(payload_root.resolve()):
        raise ValueError(f"path traversal detected: {result}")
    return result

File: packages/server/src/storage.ts
/**
 * MFUP/2 SQLite storage layer (node:sqlite).
 *
 * Each session lives in:
 *     <base_dir>/.incoming.<session_id>/state.sqlite
 *     <base_dir>/.incoming.<session_id>/payload/...
 *
 * Mirrors server/mfup-core/mfup_core/storage.py: same schema, same pragmas,
 * same batch semantics. node:sqlite is synchronous (like better-sqlite3);
 * batched transactions keep each event-loop hold in the microseconds.
 */

import { DatabaseSync } from "node:sqlite";
import * as fs from "node:fs";
import * as path from "node:path";

import { NodeKind, NodeStatus, SessionState, ROOT_NODE_ID } from "./protocol.js";

const _SCHEMA = `
CREATE TABLE IF NOT EXISTS sessions (
    session_id   TEXT PRIMARY KEY,
    resume_token TEXT NOT NULL,
    epoch        INTEGER NOT NULL DEFAULT 1,
    state        TEXT NOT NULL DEFAULT 'active',
    target_dir   TEXT NOT NULL DEFAULT '.',
    expires_at   TEXT NOT NULL,
    created_at   TEXT NOT NULL,
    updated_at   TEXT NOT NULL
);

CREATE TABLE IF NOT EXISTS nodes (
    node_id    INTEGER PRIMARY KEY,
    parent_id  INTEGER NOT NULL,
    kind       TEXT NOT NULL,
    name       TEXT NOT NULL,
    size       INTEGER,
    mtime_ms   INTEGER,
    status     TEXT NOT NULL DEFAULT 'open'
);

CREATE TABLE IF NOT EXISTS files (
    node_id          INTEGER PRIMARY KEY REFERENCES nodes(node_id),
    accepted_offset  INTEGER NOT NULL DEFAULT 0,
    final_size       INTEGER,
    checksum_state   BLOB,
    local_tmp_path   TEXT
);

CREATE TABLE IF NOT EXISTS pruned (
    node_id  INTEGER PRIMARY KEY
);

CREATE TABLE IF NOT EXISTS rejected (
    node_id  INTEGER PRIMARY KEY,
    code     TEXT NOT NULL,
    reason   TEXT NOT NULL
);

CREATE TABLE IF NOT EXISTS root_summary (
    id               INTEGER PRIMARY KEY CHECK (id = 1),
    scan_done_units  INTEGER NOT NULL DEFAULT 0,
    scan_est_units   INTEGER NOT NULL DEFAULT 0,
    body_done_bytes  INTEGER NOT NULL DEFAULT 0,
    body_est_bytes   INTEGER NOT NULL DEFAULT 0,
    sealed           INTEGER NOT NULL DEFAULT 0
);

INSERT OR IGNORE INTO root_summary (id) VALUES (1);
`;

export interface SessionRow {
  session_id: string;
  resume_token: string;
  epoch: number;
  state: string;
  target_dir: string;
  expires_at: string;
  created_at: string;
  updated_at: string;
  meta_json: string | null;
  auth_json: string | null;
}

export interface NodeRow {
  node_id: number;
  parent_id: number;
  kind: "dir" | "file";
  name: string;
  size: number | null;
  mtime_ms: number | null;
  status: string;
}

export interface FileRow {
  node_id: number;
  accepted_offset: number;
  final_size: number | null;
  checksum_state: Uint8Array | null;
  local_tmp_path: string | null;
}

export interface RootSummaryData {
  scan_done_units: number;
  scan_est_units: number;
  body_done_bytes: number;
  body_est_bytes: number;
  sealed: boolean;
}

function nowIso(): string {
  return new Date().toISOString();
}

/**
 * Thin wrapper around a per-session SQLite database.
 *
 * SQLite autocommits each statement outside an explicit transaction, so the
 * Python `_maybe_commit` dance collapses to: begin_batch() opens a
 * transaction, end_batch()/flush() commit it. For hot paths (a data POST
 * carrying thousands of frames) the batch is 10–100× cheaper for
 * many-small-files uploads (think node_modules).
 */
export class SessionDB {
  readonly path: string;
  private db: DatabaseSync;
  private batch = false;

  constructor(dbPath: string) {
    this.path = dbPath;
    this.db = new DatabaseSync(dbPath);
    this.db.exec("PRAGMA journal_mode=WAL");
    this.db.exec("PRAGMA busy_timeout=3000");
    this.db.exec("PRAGMA synchronous=NORMAL");
    this.db.exec(_SCHEMA);
    // Lightweight migration: sessions are short-lived (TTL), so additive
    // columns are applied in place instead of versioned migrations.
    for (const col of ["meta_json", "auth_json"]) {
      try {
        this.db.exec(`ALTER TABLE sessions ADD COLUMN ${col} TEXT`);
      } catch {
        /* column already exists */
      }
    }
  }

  close(): void {
    if (this.batch) {
      try {
        this.db.exec("COMMIT");
      } catch {
        /* not in a transaction */
      }
      this.batch = false;
    }
    this.db.close();
  }

  // -- batching -------------------------------------------------------------

  /** Enter deferred-commit mode (idempotent). */
  beginBatch(): void {
    if (!this.batch) {
      this.db.exec("BEGIN");
      this.batch = true;
    }
  }

  /** Commit accumulated work and leave deferred-commit mode. */
  endBatch(): void {
    if (this.batch) {
      this.db.exec("COMMIT");
      this.batch = false;
    }
  }

  /** Commit accumulated work but stay in the current mode. */
  flush(): void {
    if (this.batch) {
      this.db.exec("COMMIT");
      this.db.exec("BEGIN");
    }
  }

  // -- session --------------------------------------------------------------

  initSession(
    sessionId: string,
    resumeToken: string,
    expiresAt: string,
    targetDir = ".",
    metaJson: string | null = null,
  ): void {
    const now = nowIso();
    // Explicit column list: the table gains additive columns over time.
    this.db
      .prepare(
        "INSERT OR REPLACE INTO sessions " +
          "(session_id, resume_token, epoch, state, target_dir, expires_at," +
          " created_at, updated_at, meta_json) " +
          "VALUES (?,?,1,?,?,?,?,?,?)",
      )
      .run(sessionId, resumeToken, SessionState.ACTIVE, targetDir, expiresAt, now, now, metaJson);
  }

  /** Persist authorize-hook constraints (quotas + JSON-able context) so
   * they survive restarts / recovery. */
  setAuthJson(authJson: string): void {
    this.db.prepare("UPDATE sessions SET auth_json=?").run(authJson);
  }

  getSession(): SessionRow | null {
    const row = this.db.prepare("SELECT * FROM sessions LIMIT 1").get();
    return (row as unknown as SessionRow) ?? null;
  }

  getTargetDir(): string {
    const row = this.db.prepare("SELECT target_dir FROM sessions LIMIT 1").get() as
      | { target_dir: string }
      | undefined;
    return row ? row.target_dir : ".";
  }

  setState(state: SessionState): void {
    this.db.prepare("UPDATE sessions SET state=?, updated_at=?").run(state, nowIso());
  }

  getState(): SessionState {
    const row = this.db.prepare("SELECT state FROM sessions LIMIT 1").get() as
      | { state: string }
      | undefined;
    return row ? (row.state as SessionState) : SessionState.ACTIVE;
  }

  incrementEpoch(): number {
    this.db.prepare("UPDATE sessions SET epoch = epoch + 1, updated_at = ?").run(nowIso());
    const row = this.db.prepare("SELECT epoch FROM sessions LIMIT 1").get() as { epoch: number };
    return row.epoch;
  }

  getEpoch(): number {
    const row = this.db.prepare("SELECT epoch FROM sessions LIMIT 1").get() as
      | { epoch: number }
      | undefined;
    return row ? row.epoch : 0;
  }

  getExpiresAt(): string {
    const row = this.db.prepare("SELECT expires_at FROM sessions LIMIT 1").get() as {
      expires_at: string;
    };
    return row.expires_at;
  }

  setExpiresAt(expiresAt: string): void {
    this.db.prepare("UPDATE sessions SET expires_at=?, updated_at=?").run(expiresAt, nowIso());
  }

  // -- nodes ----------------------------------------------------------------

  /**
   * Upsert a node. Returns true iff a NEW files-row was created (i.e. this is
   * the first time we see this FILE node) — used by the caller to maintain
   * the file-count quota without SQL aggregates.
   */
  upsertNode(
    nodeId: number,
    parentId: number,
    kind: NodeKind,
    name: string,
    size: number | null = null,
    mtimeMs: number | null = null,
  ): boolean {
    const kindStr = kind === NodeKind.DIR ? "dir" : "file";
    this.db
      .prepare("INSERT OR REPLACE INTO nodes VALUES (?,?,?,?,?,?,?)")
      .run(nodeId, parentId, kindStr, name, size, mtimeMs, NodeStatus.OPEN);
    let newFile = false;
    if (kind === NodeKind.FILE) {
      const res = this.db
        .prepare("INSERT OR IGNORE INTO files (node_id, accepted_offset) VALUES (?, 0)")
        .run(nodeId);
      newFile = res.changes === 1;
    }
    return newFile;
  }

  getNode(nodeId: number): NodeRow | null {
    const row = this.db.prepare("SELECT * FROM nodes WHERE node_id=?").get(nodeId);
    return (row as unknown as NodeRow) ?? null;
  }

  setNodeStatus(nodeId: number, status: NodeStatus): void {
    this.db.prepare("UPDATE nodes SET status=? WHERE node_id=?").run(status, nodeId);
  }

  // -- files ----------------------------------------------------------------

  getFile(nodeId: number): FileRow | null {
    const row = this.db.prepare("SELECT * FROM files WHERE node_id=?").get(nodeId);
    return (row as unknown as FileRow) ?? null;
  }

  setAcceptedOffset(nodeId: number, offset: number): void {
    this.db.prepare("UPDATE files SET accepted_offset=? WHERE node_id=?").run(offset, nodeId);
  }

  setFileFinal(nodeId: number, finalSize: number, localPath: string): void {
    this.db
      .prepare("UPDATE files SET final_size=?, local_tmp_path=? WHERE node_id=?")
      .run(finalSize, localPath, nodeId);
  }

  getAllFiles(): { node_id: number; accepted_offset: number; status: string }[] {
    return this.db
      .prepare(
        "SELECT f.node_id, f.accepted_offset, n.status " +
          "FROM files f JOIN nodes n ON f.node_id = n.node_id",
      )
      .all() as { node_id: number; accepted_offset: number; status: string }[];
  }

  // -- prune / reject -------------------------------------------------------

  addPruned(nodeId: number): void {
    this.db.prepare("INSERT OR IGNORE INTO pruned VALUES (?)").run(nodeId);
  }

  isPruned(nodeId: number): boolean {
    return this.db.prepare("SELECT 1 FROM pruned WHERE node_id=?").get(nodeId) !== undefined;
  }

  getPrunedNodes(): number[] {
    return (this.db.prepare("SELECT node_id FROM pruned").all() as { node_id: number }[]).map(
      (r) => r.node_id,
    );
  }

  addRejected(nodeId: number, code: string, reason: string): void {
    this.db.prepare("INSERT OR IGNORE INTO rejected VALUES (?,?,?)").run(nodeId, code, reason);
    this.setNodeStatus(nodeId, NodeStatus.REJECTED);
  }

  isRejected(nodeId: number): boolean {
    return this.db.prepare("SELECT 1 FROM rejected WHERE node_id=?").get(nodeId) !== undefined;
  }

  getRejectedFiles(): number[] {
    return (this.db.prepare("SELECT node_id FROM rejected").all() as { node_id: number }[]).map(
      (r) => r.node_id,
    );
  }

  // -- root summary ---------------------------------------------------------

  updateRootSummary(
    scanDone: number,
    scanEst: number,
    bodyDone: number,
    bodyEst: number,
    sealed: boolean,
  ): void {
    this.db
      .prepare(
        "UPDATE root_summary SET scan_done_units=?, scan_est_units=?, " +
          "body_done_bytes=?, body_est_bytes=?, sealed=? WHERE id=1",
      )
      .run(scanDone, scanEst, bodyDone, bodyEst, sealed ? 1 : 0);
  }

  getRootSummary(): RootSummaryData {
    const row = this.db.prepare("SELECT * FROM root_summary WHERE id=1").get() as
      | {
          scan_done_units: number;
          scan_est_units: number;
          body_done_bytes: number;
          body_est_bytes: number;
          sealed: number;
        }
      | undefined;
    if (!row) {
      return {
        scan_done_units: 0,
        scan_est_units: 0,
        body_done_bytes: 0,
        body_est_bytes: 0,
        sealed: false,
      };
    }
    return {
      scan_done_units: row.scan_done_units,
      scan_est_units: row.scan_est_units,
      body_done_bytes: row.body_done_bytes,
      body_est_bytes: row.body_est_bytes,
      sealed: !!row.sealed,
    };
  }

  // -- aggregate stats for COMMIT_OK ----------------------------------------

  /** Total nodes seen (dirs + files) — commit node-count invariant. */
  countNodes(): number {
    const row = this.db.prepare("SELECT COUNT(*) AS n FROM nodes").get() as { n: number };
    return row.n;
  }

  /** Return [fileCount, totalBytes] for all non-rejected files. */
  countCommittedFiles(): [number, number] {
    const row = this.db
      .prepare(
        "SELECT COUNT(*) AS n, COALESCE(SUM(accepted_offset), 0) AS b " +
          "FROM files f JOIN nodes n ON f.node_id = n.node_id " +
          "WHERE n.status NOT IN ('rejected', 'pruned')",
      )
      .get() as { n: number; b: number };
    return [row.n, row.b];
  }

  /** Files where accepted_offset != final_size (incomplete transfers). */
  getIncompleteFiles(): { node_id: number; accepted_offset: number; final_size: number | null }[] {
    return this.db
      .prepare(
        "SELECT f.node_id, f.accepted_offset, f.final_size " +
          "FROM files f JOIN nodes n ON f.node_id = n.node_id " +
          "WHERE n.status NOT IN ('rejected', 'pruned') " +
          "  AND (f.final_size IS NULL OR f.accepted_offset != f.final_size)",
      )
      .all() as { node_id: number; accepted_offset: number; final_size: number | null }[];
  }
}

// ---------------------------------------------------------------------------
// Staging directory helpers
// ---------------------------------------------------------------------------

export const DEFAULT_STAGING_PREFIX = ".incoming";

export function stagingDir(
  baseDir: string,
  sessionId: string,
  prefix: string = DEFAULT_STAGING_PREFIX,
): string {
  return path.join(baseDir, `${prefix}.${sessionId}`);
}

export function ensureStaging(
  baseDir: string,
  sessionId: string,
  prefix: string = DEFAULT_STAGING_PREFIX,
): string {
  const sd = stagingDir(baseDir, sessionId, prefix);
  fs.mkdirSync(path.join(sd, "payload"), { recursive: true });
  return sd;
}

export function openSessionDb(
  baseDir: string,
  sessionId: string,
  prefix: string = DEFAULT_STAGING_PREFIX,
): SessionDB {
  const sd = ensureStaging(baseDir, sessionId, prefix);
  return new SessionDB(path.join(sd, "state.sqlite"));
}

/** Reject names that could escape the payload directory. */
export function validateNodeName(name: string): void {
  if (!name) {
    throw new Error("empty node name");
  }
  if (name === "." || name === "..") {
    throw new Error(`illegal node name: ${JSON.stringify(name)}`);
  }
  if (name.includes("/") || name.includes("\\") || name.includes("\x00")) {
    throw new Error(`illegal characters in node name: ${JSON.stringify(name)}`);
  }
}

/**
 * Build the filesystem path for a node inside the payload directory.
 *
 * Walks the parent_id chain in the DB to reconstruct the relative path.
 * Validates each name component to prevent path traversal.
 */
export function resolvePayloadPath(
  baseDir: string,
  sessionId: string,
  db: SessionDB,
  nodeId: number,
  prefix: string = DEFAULT_STAGING_PREFIX,
): string {
  const parts: string[] = [];
  let curId = nodeId;
  const seen = new Set<number>();
  while (curId !== ROOT_NODE_ID) {
    if (seen.has(curId)) {
      throw new Error(`parent cycle detected at node ${curId}`);
    }
    seen.add(curId);
    const node = db.getNode(curId);
    if (node === null) {
      // A silent break here would resolve a SHORTENED path and write the
      // file into the wrong directory. Broken chains are an error.
      throw new Error(`broken parent chain: node ${curId} unknown`);
    }
    validateNodeName(node.name);
    parts.push(node.name);
    curId = node.parent_id;
  }
  parts.reverse();
  const payloadRoot = path.join(stagingDir(baseDir, sessionId, prefix), "payload");
  const result = parts.length > 0 ? path.join(payloadRoot, ...parts) : payloadRoot;
  // Defense in depth: resolved path must stay within the payload dir.
  const rel = path.relative(path.resolve(payloadRoot), path.resolve(result));
  if (rel.startsWith("..") || path.isAbsolute(rel)) {
    throw new Error(`path traversal detected: ${result}`);
  }
  return result;
}
